pgbedrock/attributes.py

```diff
@@ -194,13 +204,47 @@ def get_attribute_value(self, attribute):
         return value
 
     def is_same_password(self, value):
-        """ Convert the input value into a postgres rolname-salted md5 hash and compare
-        it with the currently stored hash """
-        if value is None:
-            return self.current_attributes.get('rolpassword') is None
+        """ Compare stored password hash to the new password.
+
+        Returns ``True`` if both ``value`` and the stored password are ``None`` (empty)
+        **or** ``value`` is a plain text password its hash matches the stored password
+        hash **or** ``value`` is a hashed password and matches the stored password hash
+        type and value. ``False`` in any other case. """
+        current_value = self.current_attributes.get('rolpassword')
+
+        if current_value == value:
+            return True
+
+        if value is None or current_value is None:
+            return False
+
+        if current_value.startswith("SCRAM-SHA-256$"):
+            if value.startswith("md5"):
+                return False
+
+            if value.startswith("SCRAM-SHA-256$"):
+                return current_value == value
+
+            hash_parts = current_value.split("$")
+
+            iterations, salt = hash_parts[1].split(":")
+            iterations = int(iterations)
+            salt = base64.b64decode(salt)
+
+            return current_value == create_scram_hash(value, salt, iterations)
+
+        if current_value.startswith("md5"):
+            if value.startswith("SCRAM-SHA-256$"):
+                return False
+
+            if value.startswith("md5"):
+                return current_value == value
+
+            return current_value == create_md5_hash(self.rolename, value)
```

Comment:
I am going to be fancy and recommend you use https://docs.python.org/3/library/hmac.html#hmac.compare_digest I.e constant time string comparison. It's completely useless because no-one would use pgbedrock to brute force / guess a PG password, but hey the more you know, the more you can show off !

Reply:
alors, @johanfleury please do test this in Staging, we don't want pgbedock to mistakenly update passwords in prod :D